src/js/tls.js

```diff
@@ -3,85 +3,109 @@
 require('crypto');
 
 var net = require('net');
+var tcp = require('tcp');
 var util = require('util');
-var EventEmitter = require('events').EventEmitter;
+/** @type tls.TlsWrap */
 var TlsWrap = native.TlsWrap;
 var TLS_CHUNK_MAX_SIZE = require('constants').TLS_CHUNK_MAX_SIZE;
 
-function TLSSocket(socket, opts) {
-  if (!(this instanceof TLSSocket))
-    return new TLSSocket(socket, opts);
-  EventEmitter.call(this);
-
-  var tlsOptions = Object.assign({
-    servername: socket.host || socket.hostname,
-    // rejectUnauthorized: false,
-  }, opts);
-
-  // TODO: currently certification is disabled
-  tlsOptions.rejectUnauthorized = false;
-
-  // handle the [ca1,ca2,...]
-  if (Array.isArray(tlsOptions.ca)) {
-    tlsOptions.ca = tlsOptions.ca.join('\n');
+function Delegate(proto, target) {
+  if (!(this instanceof Delegate)) {
+    return new Delegate(proto, target);
   }
+  this.proto = proto;
+  this.target = target;
+}
 
-  this.servername = tlsOptions.servername;
-  this.authorized = false;
-  this.authorizationError = null;
-  this._socket = new net.Socket(socket);
-  this._writev = [];
-  this._ended = false;
+Delegate.prototype.method = function method(name) {
+  var proto = this.proto;
+  var target = this.target;
 
-  // Just a documented property to make secure sockets
-  // distinguishable from regular ones.
-  this.encrypted = false;
-  this._socket.on('error', this._tlsError.bind(this));
-  this._socket.on('connect', this.onsocket.bind(this));
-  this._socket.on('data', this.onsocketdata.bind(this));
+  proto[name] = this[target][name].bind(this[target]);
```

Comment:
Actually bond function still lost the function name :(

Reply:
How does just inherit from `net.Socket` like https://github.com/nodejs/node/blob/master/lib/_tls_wrap.js#L290-L341? I think this implementation is that we should follow yet :)

Reply:
`tls.TLSSocket` does inherit from `net.Socket` in this implementation, while `TLSHandle` doesn't. What `TLSHandle` do is to implement an interface that `net.Socket` interacts with.